Add a new serialized type: STNumber

thejohnfreeman opened the pull request:
This type should let objects and transactions contain multiple fields for quantities of XRP, IOU, or [MPT](https://github.com/XRPLF/rippled/pull/5108) without duplicating information about the "issue" (represented by `STIssue`). It is a straightforward serialization of our internal `Number` type that uniformly represents those quantities.

dangell7 commented:
I like this. I had all sorts of fun when doing `sfQuantity` on the options proposal. I think I just ended up using an i32. Some things I would be curious about. 

1. Will there be a `getFieldNumber(sfQuantity)`? (Assuming yes)
2. Also how can one multiply / divide / add / subtract an STNumber with an STAmount. That might be good to show in tests if its not straight forward. For example I have 100 USD and I want to multiply by 100. 

I ended up with `STAmount const totalValue = STAmount(strikePrice.issue(), (strikePrice.mantissa() * quantity));` but maybe I took a wrong turn. 

Thank you

gregtatcam commented on src/libxrpl/protocol/Serializer.cpp:
```diff
@@ -71,6 +72,32 @@ Serializer::add64(std::uint64_t i)
+    static_assert(SIZE == 4);
+    mData.resize(ret + SIZE);
+    unsigned char* p = &mData.back() + 1 - SIZE;
+    boost::endian::store_big_s32(p, i);
```
Why not use the same serialization approach as in other methods? I.e. `push_back/shift`?

Speedtraders48 replied:
Because the hour and time ratio are not the same.

gregtatcam commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,63 @@
+
+namespace ripple {
+
+class STNumber final : public STBase
```
Do we need to inherit from `CountedObject`? Most `ST*` types do.

Speedtraders48 replied:
Yes and also make it perfect

gregtatcam commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,63 @@
+
+namespace ripple {
+
+class STNumber final : public STBase
```
Why contain and not inherit from `Number`? All arithmetics then just works.

Speedtraders48 replied:
From the decimal point but it is perfect

gregtatcam commented on src/test/protocol/STNumber_test.cpp:
```diff
@@ -0,0 +1,61 @@
+    void
+    run() override
+    {
+        {
```
Does it make sense to have this block of code in a separate function, for instance `testNumber`. And have `run` call `testNumber`? We'll probably be adding more tests for `STNumber`.

gregtatcam replied:
I guess we don't want to have the arithmetics work between not compatible types.

thejohnfreeman replied:
I split tests out into separate functions at the point it becomes meaningful.

thejohnfreeman commented:
@dangell7 I added `getFieldNumber`. I generally add methods as-needed, and the `getField*` methods seem incomplete (there is no `getFieldIssue`, for example). Let me know if there are others you want.

I added a test to demonstrate multiplication. The general pattern is to convert everything to `Number`, compute your arithmetic, and then reconstruct final types like `STAmount`. We can add arithmetic helpers for those final types where it makes sense. I'm a little disappointed that conversion from `STAmount` to `Number` must first pass through constructing either an `XRPAmount` or an `IOUAmount`, though I guess it is inlined.

gregtatcam commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,66 @@
+    STNumber(SerialIter& sit, SField const& name);
+
+    using Number::operator=;
+
```
I think should add `using value_type = Number;`. It's used by `STObject::operator[]`, etc.

gregtatcam commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,66 @@
+    Number const&
+    value() const;
+    void
+    setValue(Number const& v);
```
Is this needed? Why not use assignment operator?

thejohnfreeman replied:
I think it's for `STObject::setFieldUsingSetValue`.

gregtatcam commented on src/test/protocol/Serializer_test.cpp:
```diff
@@ -0,0 +1,51 @@
+    void
+    run() override
+    {
+        std::initializer_list<std::int64_t> const values = {
```
Does it also make sense to add `int32` tests?

gregtatcam commented on src/test/protocol/STNumber_test.cpp:
```diff
@@ -0,0 +1,72 @@
+            BEAST_EXPECT(stnum.value() == Number{0});
+        }
+
+        std::initializer_list<std::int64_t> const values = {
```
Should add `int32` tests?

gregtatcam commented on src/test/protocol/STNumber_test.cpp:
```diff
@@ -0,0 +1,72 @@
+            IOUAmount totalValue{iouValue * factor};
+            STAmount const totalAmount{totalValue, strikePrice.issue()};
+            BEAST_EXPECT(totalAmount == Number{10'000});
+        }
```
Should we test arithmetics with some large/small fractional values to make sure `STAmount` and `STNumber` produce the same results?

thejohnfreeman replied:
I welcome all tests written by others, but I don't want to spend my time designing those tests right now.

gregtatcam commented on src/libxrpl/protocol/STNumber.cpp:
```diff
@@ -0,0 +1,104 @@
+
+Json::Value STNumber::getJson(JsonOptions) const
+{
+    // TODO: come up with a string representation.
```
How about 

     elem[jss::value] = getText();

thejohnfreeman replied:
That will give us things like `{ "sfFieldName": { "value": "1234567" }}` instead of just `{ "sfFieldName": "1234567" }`. I think `to_string` might just be the correct representation. I'll remove the comment.

thejohnfreeman commented on src/libxrpl/protocol/Serializer.cpp:
```diff
@@ -71,6 +72,32 @@ Serializer::add64(std::uint64_t i)
 }
 
+int
+Serializer::addi32(std::int32_t i)
```
My thinking was grouping signed vs unsigned before bit length, but I'm fine either way.

thejohnfreeman commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,66 @@
+
+namespace ripple {
+
+class STNumber final : public STBase,
```
I didn't put much thought into it. Most of the `ST` classes are final. We can always remove it when it's time to inherit, but I'll remove it now anyways.

thejohnfreeman commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,66 @@
+    bool
+    isEquivalent(STBase const& t) const override;
+    bool
+    isDefault() const override;
```
Won't it just use the existing overload for `Number`?

gregtatcam replied:
It's ambiguous because both `Number` and `STBase` have it overloaded.

thejohnfreeman commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -0,0 +1,66 @@
+public:
+    STNumber() = default;
+    explicit STNumber(SField const& n, Number const& v = Number());
+    STNumber(SerialIter& sit, SField const& name);
```
I'm hesitant to add this before we have any callers that need it.

gregtatcam reviewed (approved):
👍 LGTM
I left a couple of minor comments to consider.

HowardHinnant commented:
One thing that bothers me is the ability to destruct a `STNumber` via a reference or pointer to a `Number`.  This would call `Number`'s non-virtual destructor and not call `STNumber`'s destructor.

That error can be prevented at compile-time with this refactor: https://github.com/HowardHinnant/rippled/commit/05ecfc6ef279475d695b065a35ae82476ba65f13

Only very lightly tested.  So if accepted, this does need further scrutiny.

gregtatcam commented:
> One thing that bothers me is the ability to destruct a `STNumber` via a reference or pointer to a `Number`. This would call `Number`'s non-virtual destructor and not call `STNumber`'s destructor.
> 
> That error can be prevented at compile-time with this refactor: [HowardHinnant@05ecfc6](https://github.com/HowardHinnant/rippled/commit/05ecfc6ef279475d695b065a35ae82476ba65f13)
> 
> Only very lightly tested. So if accepted, this does need further scrutiny.

@HowardHinnant Why not adding a virtual destructor? If `STNumber` is a subclass of `Number` then the math just works.

HowardHinnant commented:
> @HowardHinnant Why not adding a virtual destructor? If `STNumber` is a subclass of `Number` then the math just works.

The math also just works with the way I refactored it.  Note that I did not have to change any use cases of `STNumber`, in code or in tests.

Adding a virtual destructor to `Number` adds an expense to every client of `Number` whether they use it or not.  My refactor adds no expense to `Number` at all.  It does not touch the declaration or implementation of `Number`. That makes it a cleaner separation of duties.  And the resulting API that clients see is identical, minus the ability to implicitly convert `STNumber*` to `Number*` (and `STNumber&` to `Number&`), which is both unneeded and undesirable.

ximinez commented:
Waiting on 2.3.0 release

thejohnfreeman commented:
```
Add a new serialized type: STNumber (#5121)

`STNumber` lets objects and transactions contain multiple fields for
quantities of XRP, IOU, or MPT without duplicating information about the
"issue" (represented by `STIssue`). It is a straightforward serialization of
the `Number` type that uniformly represents those quantities.
```
